kr_mesh_visualization/src/mesh_visualization.cpp

```diff
@@ -1,83 +1,100 @@
+    this->get_parameter("color/a", color_a);
+    this->get_parameter("scale/x", scale_x);
+    this->get_parameter("scale/y", scale_y);
+    this->get_parameter("scale/z", scale_z);
+    this->get_parameter("new_frame_id", new_frame_id);
 
-  if(msg->getDataType() == "geometry_msgs/PoseStamped")
+    using namespace std::placeholders;
+    pub_vis_ = this->create_publisher<visualization_msgs::msg::Marker>("~/robot", 10);
+    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>("~/input", 10,
+                std::bind(&MeshVisualizationNode::odom_callback, this, _1));
+    
+
+    /* Uncomment any of the below subscribers to support another message type */
+    // sub_pwcs_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
+    //     "~/input", 10, std::bind(&MeshVisualizationNode::callback_pose_with_covariance_stamped, this, _1));
+    // sub_ps_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
+    //     "~/input", 10, std::bind(&MeshVisualizationNode::callback_pose_stamped, this, _1));
+  }
+
+ protected:
+  std::string mesh_resource, new_frame_id;
+  rclcpp::Publisher<visualization_msgs::msg::Marker>::SharedPtr pub_vis_;
+  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_ps_;
+  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr sub_pwcs_;
+  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
+  double color_r, color_g, color_b, color_a;
+  double scale_x, scale_y, scale_z;
+
+  void publishMarker(const std::string &frame_id, const geometry_msgs::msg::Pose &pose)
   {
-    auto pose_msg = msg->instantiate<geometry_msgs::PoseStamped>();
-    publishMarker(pose_msg->header.frame_id, pose_msg->pose);
+    auto marker = visualization_msgs::msg::Marker();
+
+    marker.header.frame_id = (new_frame_id == "") ? frame_id : new_frame_id;
+    marker.header.stamp = rclcpp::Time(0, 0);
+    marker.ns = this->get_name();
+    marker.id = 0;
+    marker.type = visualization_msgs::msg::Marker::MESH_RESOURCE;
+    marker.action = visualization_msgs::msg::Marker::ADD;
+    marker.pose = pose;
+    marker.scale.x = scale_x;
+    marker.scale.y = scale_y;
+    marker.scale.z = scale_z;
+    marker.color.a = color_a;
+    marker.color.r = color_r;
+    marker.color.g = color_g;
+    marker.color.b = color_b;
+    marker.mesh_resource = mesh_resource;
+    pub_vis_->publish(marker);
   }
-  else if(msg->getDataType() == "geometry_msgs/PoseWithCovarianceStamped")
+
+  void callback_pose_stamped(const geometry_msgs::msg::PoseStamped::SharedPtr msg)
   {
-    auto pose_msg = msg->instantiate<geometry_msgs::PoseWithCovarianceStamped>();
-    publishMarker(pose_msg->header.frame_id, pose_msg->pose.pose);
+    publishMarker(msg->header.frame_id, msg->pose);
   }
-  else if(msg->getDataType() == "nav_msgs/Odometry")
+
+  void callback_pose_with_covariance_stamped(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg)
```

Comment:
Comment the callbacks that we are not using

Reply:
Okay, makes sense. Will do.